### ISSUE TITLE
[BUG] AltStore compatible source throws ("E downloadURL:String or downloadURLs:[[PlatformURL]] key required." or "CoreData error") in SideStore

### ISSUE THREAD

**Issue by jgottula** [SideStore/SideStore#735]
### Describe the bug

When attempting to add a new source, the source cannot be added, and the following error message appears:
```
The data couldn’t be read because it isn’t in the correct format.

E downloadURL:String or downloadURLs:[[Platform:URL]] key required.
```

If a source was previously successfully added, but then at some future point the contents of that source URL change (in the same manner that would have caused the error-while-adding-new-source shown above), the user instead sees a cryptic CoreData error when SideStore attempts to refresh its sources; and some (or possibly all?) sources will not actually update (irrespective of whether those other sources themselves have any actual issues) until all sources with the problem are removed from SideStore. (This particular behavior is a pretty big deal.)

### Instructions to reproduce

Add a source, in which one or more of the apps described by the file has a "versions" key but which does NOT also have a "downloadURL" key at its top level (i.e. outside of / separate from the "versions" array).

Having a "versions" array (with separate versions having their own respective "downloadURL" keys), and not having an additional/redundant top-level "downloadURL" key for the entire app, is something which is considered legal by AltStore, but which SideStore cannot handle and considers to be an error fatal to the validity of the entire source JSON file.

SideStore always expects there to be a "downloadURL" key at the app top-level, regardless of the presence of a "versions" array. (AltStore, by contrast, essentially uses the "downloadURL" key of the latest version from the "versions" array as the app's top-level download URL, when a "versions" array is present.)

---

I have a very helpful real-world case that clearly demonstrates the issue and how to reproduce it: the AltStore source file for the uYouEnhanced app.

In commit https://github.com/arichornlover/arichornlover.github.io/commit/7bd125b962342f5214e733fbef12462357e4aa2f, the only change of relevance that was made to the file is that the top-level app "downloadURL" key was removed. (There are two versions in its "versions" array, each of which has its own valid "downloadURL" key.)

If you add this source file, as it existed just before that commit, to SideStore, it works. But if you add the source file, as it exists with that commit applied, to SideStore, it no longer works and it produces the error message(s) and effects described earlier.

WORKS (c5b7240): https://raw.githubusercontent.com/arichornlover/arichornlover.github.io/c5b7240/apps.json
BROKEN (7bd125b): https://raw.githubusercontent.com/arichornlover/arichornlover.github.io/7bd125b/apps.json

### What version of SideStore are you using?

0.5.8

### Other info

This seems to be by far the most common AltStore source file compatibility issue, so it's probably worth attempting to fix it. (Even if fully addressing all incompatibilities with AltStore's source file format is a "later" kind of thing.)

The current public GitHub source code for AltStore does show the different way that the downloadURL situation is handled. All of the relevant stuff is in class StoreApp in AltStoreCore/Model/StoreApp.swift.

AltStore has logic that handles cases where there is a "versions" array and no top-level "downloadURL" key; whereas SideStore unconditionally expects a top-level "downloadURL" key, regardless of the presence of a "versions" array.

[SideStore](https://github.com/SideStore/SideStore/blob/develop/AltStoreCore/Model/StoreApp.swift)
[AltStore](https://github.com/altstoreio/AltStore/blob/main/AltStoreCore/Model/StoreApp.swift)

I would like to submit a PR for this myself, but unfortunately I'm not a Swift guy and I also don't have an Xcode environment accessible to me, so that unfortunately makes things difficult. But hopefully I can at the very least provide enough details here to make it relatively straightforward to craft a fix for this issue without tremendously much effort.

Relevant: #314, #564

**Comment by jgottula** [comment]
Was just testing this again using the 0.5.10 release. (I failed to notice the existence of 0.5.9... guess I oughta turn on release notifications in GitHub.)

Problem still exists in 0.5.10, merely causing a different error message now.
```
No value associated with key CodingKeys(string value: "version", intValue: nil) ("version").
```
So yeah, I guess there actually was more to it than just the decoding logic that got fixed up in #807.

I haven't had a chance to test this with the 0.6.0 alpha or a nightly build (I don't have a scratch device and I'm reluctant to risk screwing anything up on my real in-use devices), but I'm guessing that things are probably looking better over there.

**Comment by mahee96** [comment]
yes the logic is to use the downloadURL of the first app in the versions list, so if both downloadURL in top level is absent and there are no versions, then only this error should be thrown.

check recent nightly, if it is still broken wait for PR https://github.com/SideStore/SideStore/pull/866 to be merged, it is undergoing testing phase rn.

**Comment by mahee96** [comment]
recently nightly contains the fix, the downloadURL error won't happen if it is defined at 1. apps.first.versions.first else if apps.first.downloadURL (backward compatibility).
